src/notification.c

```diff
@@ -234,129 +234,184 @@ check_value_type (const char *key,
-                   XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
-                   "%s not a priority", g_variant_get_string (value, NULL));
-      return FALSE;
-    }
+  priority = g_variant_get_string (value, NULL);
+
+  if (g_strv_contains (priorities, priority))
+    g_variant_builder_add (builder, "{sv}", "priority", value);
+  else
+    g_debug ("Unsupported priority %s filtered from notification", priority);
 
   return TRUE;
 }
 
 static gboolean
-check_button (GVariant *button,
-              GError **error)
+parse_button (GVariantBuilder  *builder,
+              GVariant         *button,
+              GError          **error)
 {
   int i;
-  gboolean has_label = FALSE;
-  gboolean has_action = FALSE;
+  g_autoptr(GVariant) label = NULL;
+  g_autoptr(GVariant) action = NULL;
+  g_autoptr(GVariant) target = NULL;
 
   for (i = 0; i < g_variant_n_children (button); i++)
     {
       const char *key;
       g_autoptr(GVariant) value = NULL;
 
       g_variant_get_child (button, i, "{&sv}", &key, &value);
+
       if (strcmp (key, "label") == 0)
-        has_label = TRUE;
+        {
+          if (!check_value_type (key, value, G_VARIANT_TYPE_STRING, error))
+            return FALSE;
+          label = g_steal_pointer (&value);
+        }
       else if (strcmp (key, "action") == 0)
-        has_action = TRUE;
+        {
+          if (!check_value_type (key, value, G_VARIANT_TYPE_STRING, error))
+            return FALSE;
+          action = g_steal_pointer (&value);
+        }
       else if (strcmp (key, "target") == 0)
-        ;
+        {
+          target = g_steal_pointer (&value);
+        }
       else
         {
-          g_set_error (error,
-                       XDG_DESKTOP_PORTAL_ERROR,
-                       XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
-                       "%s not valid key", key);
-          return FALSE;
+          g_debug ("Unsupported button property %s filtered from notification", key);
         }
     }
 
-  if (!has_label)
+  if (label && action)
     {
-      g_set_error_literal (error,
-                           XDG_DESKTOP_PORTAL_ERROR,
-                           XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
-                           "label key is missing");
-      return FALSE;
-    }
+      g_variant_builder_open (builder, G_VARIANT_TYPE ("a{sv}"));
 
-  if (!has_action)
+      g_variant_builder_add (builder, "{sv}", "label", label);
+      g_variant_builder_add (builder, "{sv}", "action", action);
+      if (target)
+        g_variant_builder_add (builder, "{sv}", "target", target);
+
+      g_variant_builder_close (builder);
+    }
+  else
     {
-      g_set_error_literal (error,
-                           XDG_DESKTOP_PORTAL_ERROR,
-                           XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
-                           "action key is missing");
-      return FALSE;
+      g_debug ("Notification button needs a label and action property");
```

Comment:
shouldn't this `return FALSE` here? after all, there was an array entry for a button that didn't result in a button.